rclc/include/rclc/executor_handle.h

```diff
@@ -45,29 +45,62 @@ typedef enum
   ALWAYS
 } rclc_executor_handle_invocation_t;
 
+
 /// Type defintion for callback function.
 typedef void (* rclc_callback_t)(const void *);
 
+/// Type defintion for client callback function
+/// - request message
+/// - request id
+/// - response message
+typedef void (* rclc_service_callback_t)(const void *, rmw_request_id_t *, void *);
+
+/// Type defintion for client callback function
+/// - response message
+/// - request id
+typedef void (* rclc_client_callback_t)(const void *, rmw_request_id_t *);
+
+
 /// Container for a handle.
 typedef struct
 {
   /// Type of handle
   rclc_executor_handle_type_t type;
-  /// When to execute callback
+  /// Invocation type determines when to execute the callback
   rclc_executor_handle_invocation_t invocation;
+  /// Pointer to the handle
   union {
-    /// Storage of subscription pointer
     rcl_subscription_t * subscription;
-    /// Storage of timer pointer
     rcl_timer_t * timer;
-    // rcl_service_t
-    // rcl_client_t
-    // rcl_guard_condition_t
+    rcl_client_t * client;
+    rcl_service_t * service;
+    rcl_guard_condition_t * gc;
   };
   /// Storage of data, which holds the message of a subscription, service, etc.
+  /// subscription: ptr to message
+  /// service: ptr to request message
   void * data;
+
+  /// request-id only for type service/client request/response
+  rmw_request_id_t req_id;
+
+  /// only for service - ptr to response message
+  void * data_response_msg;
+
+  // TODO(jst3si) new type to be stored as data for
+  //              service/client objects
+  //              look at memory allocation for this struct!
+  // struct {
+  //   void * request_msg
+  //   void * response_msg
+  //   rmw_request_id_t req_id;
+  //} rclc_service_data_type_t
+
   /// Storage for callback for subscription
+  /// TODO(jst3si) use union
   rclc_callback_t callback;
+  rclc_service_callback_t service_callback;
```

Comment:
These callbacks could be unified in anonymous unions because they are not going to coexists, what do you think @JanStaschulat ? Something like:

```c
union{
  rclc_callback_t callback;
  rclc_service_callback_t service_callback;
  rclc_client_callback_t client_callback;
}
```

Reply:
Yes, good idea.